rmf_fleet_adapter/src/rmf_fleet_adapter/events/internal_ChopeNegotiator.hpp

```diff
@@ -0,0 +1,350 @@
+            self->_selected_waitpoint_cb(self->_waitpoints[self->_final_allocated_destination.value()->
+              satisfies_alternative]);
+            RCLCPP_INFO(
+                self->_context->node()->get_logger(), "chope: Robot %s is being asked to proceed to a to waitpoint",
+                  self->_context->name().c_str());
+          }
+        });
+
+    RCLCPP_INFO(context->node()->get_logger(),
+      "Sending chope request");
+    negotiator->make_request(same_map);
+    return negotiator;
+  }
+
+ChopeNodeNegotiator()
+  {
+
+  }
+private:
+  enum class ReservationState
+  {
+    Pending=0,
+    Requested=1,
+    RecievedResponceProceedWaitPoint=2,
+    RecievedResponseProceedImmediate=3
+  };
+
+
+
+  void make_request(bool only_same_map)
+  {
+    auto current_location = _context->location();
+    auto graph = _context->navigation_graph();
+    if (current_location.size() == 0)
+    {
+      //unable to get location. We should return some form of error stste.
+      RCLCPP_ERROR(
+        _context->node()->get_logger(),
+        "Robot [%s] can't get location",
+        _context->requester_id().c_str());
+      return;
+    }
+
+    RCLCPP_INFO(
+      _context->node()->get_logger(),
+      "Selecting a new go_to_place location from [%lu] choices for robot [%s]",
+      _goals.size(),
+      _context->requester_id().c_str());
+
+    if (_current_reservation_state == ReservationState::Pending)
+    {
+      // Select node
+      rmf_chope_msgs::msg::FlexibleTimeRequest ftr;
+      ftr.header.robot_name = _context->name();
+      ftr.header.fleet_name = _context->group();
+      ftr.header.request_id = _reservation_id;
+
+      auto lowest_cost = std::numeric_limits<double>::infinity();
+      std::optional<std::size_t> selected_idx;
+      for (std::size_t i = 0; i < _goals.size(); ++i)
+      {
+        const auto wp_idx = _goals[i].waypoint();
+        const auto& wp = graph.get_waypoint(wp_idx);
+
+        if (only_same_map)
+        {
+
+          // Check if same map. If not don't consider location. This is to ensure
+          // the robot does not try to board a lift.
+          if (wp.get_map_name() != _context->map())
+          {
+            RCLCPP_INFO(
+              _context->node()->get_logger(),
+              "Skipping [%lu] as it is on map [%s] but robot is on map [%s].",
+              wp_idx,
+              wp.get_map_name().c_str(),
+              _context->map().c_str());
+            continue;
+          }
+        }
+
+        // Find distance to said point
+        auto result =
+          _context->planner()->quickest_path(current_location, wp_idx);
+        if (result.has_value())
+        {
+          RCLCPP_INFO(
+            _context->node()->get_logger(),
+            "Got distance from [%lu] as %f",
+            wp_idx,
+            result->cost());
+
+          if (result->cost() < lowest_cost)
+          {
+            selected_idx = i;
+            lowest_cost = result->cost();
+          }
```

Comment:
I noticed that before this PR there was only one call of `quickest_path` in the whole `rmf_ros2` codebase, while now there are 4 and all the new ones seem to do one flavor of "iterates through waypoints, get quickest path, select minimum or sort by distance), it is quite a fair bit of duplicated code

Reply:
There's some redundancy, but the context for each use case is different enough that I don't think we stand to gain much by funneling it all into one function. Each has slight nuances that we would need to cram into the arguments of the single function. I think we should leave this as-is and be more mindful of hygiene in general as we develop the next generation.